confgenerator/jvm.go

```diff
@@ -0,0 +1,98 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package confgenerator
+
+import (
+	"fmt"
+	"path/filepath"
+	"runtime"
+
+	"github.com/GoogleCloudPlatform/ops-agent/confgenerator/otel"
+	"github.com/kardianos/osext"
+)
+
+type MetricsReceiverJVM struct {
+	ConfigComponent `yaml:",inline"`
+
+	MetricsReceiverShared `yaml:",inline"`
+
+	Endpoint string `yaml:"endpoint" validate:"omitempty,url"`
+	Username string `yaml:"username"`
+	Password string `yaml:"password"`
+}
+
+const defaultJVMEndpoint = "localhost:9999"
+
+func (r MetricsReceiverJVM) Type() string {
+	return "jvm"
+}
+
+func (r MetricsReceiverJVM) Pipelines() []otel.Pipeline {
+	if r.Endpoint == "" {
+		r.Endpoint = defaultJVMEndpoint
+	}
+
+	jarPath, err := findJarPath()
+	if err != nil {
+		_ = err // TODO(djaglowski) return nil, fmt.Errorf("create jvmreceiver: %v, err)
+	}
+
+	config := map[string]interface{}{
+		"target_system":       "jvm",
+		"collection_interval": r.CollectionIntervalString(),
+		"endpoint":            r.Endpoint,
+		"jar_path":            jarPath,
+	}
+
+	// Only set the username & password fields if provided
+	if r.Username != "" {
+		config["username"] = r.Username
+	}
+	if r.Password != "" {
+		config["password"] = r.Password
+	}
+
+	return []otel.Pipeline{{
+		Receiver: otel.Component{
+			Type:   "jmx",
+			Config: config,
+		},
+		Processors: []otel.Component{
+			otel.NormalizeSums(),
+			otel.MetricsTransform(
+				otel.AddPrefix("workload.googleapis.com"),
+			),
+		},
+	}}
+}
+
+func findJarPath() (string, error) {
+	jarName := "opentelemetry-java-contrib-jmx-metrics-1.0.0-alpha.jar"
```

Comment:
Ideally we aren't encoding the version number of the JAR chosen here, that should be an integration-time decision.